desc/singularities.py

```diff
@@ -0,0 +1,733 @@
+        # don't contribute to other eval pts
+        k = kernel(
+            {key: val for key, val in eval_data.items() if key in keys},
+            src_data_polar,
+            diag=True,
+        ).reshape((eval_grid.num_nodes, kernel.ndim))
+
+        dS = (v[i] * src_data_polar["|e_theta x e_zeta|"])[:, None]
+        fi = k * dS
+        return fi
+
+    def polar_pt_loop(i, f):
+        # this calculates the effect at a single evaluation point, from all others
+        # in a single field period. loop this to get all pts
+        f_temp = polar_pt_vmap(i)
+        return f + f_temp.reshape((eval_grid.num_nodes, kernel.ndim))
+
+    f = jnp.zeros((eval_grid.num_nodes, kernel.ndim))
+    # vmap found more efficient than fori_loop, esp on gpu, but uses more memory
+    if loop:
+        f = fori_loop(0, v.size, polar_pt_loop, f)
+    else:
+        f = vmap(polar_pt_vmap)(jnp.arange(v.size)).sum(axis=0)
+
+    # we sum distance vectors, so they need to be in xyz for that to work
+    # but then need to convert vectors back to rpz
+    if kernel.ndim == 3:
+        f = xyz2rpz_vec(f, phi=eval_data["zeta"])
+
+    return f
+
+
+def singular_integral(
+    eval_data,
+    eval_grid,
+    src_data,
+    src_grid,
+    kernel,
+    interpolator,
+    loop=False,
+):
+    """Evaluate a singular integral transform on a surface.
+
+    eg f(θ, ζ) = ∫ ∫ K(θ, ζ, θ', ζ') g(θ', ζ') dθ' dζ'
+
+    Uses method by Malhotra et. al. [1]_ [2]_
+
+    Parameters
+    ----------
+    eval_data : dict
+        Dictionary of data at evaluation points. Keys should be those required by
+        kernel as kernel.keys. Vector data should be in rpz basis.
+    eval_grid : Grid
+        Points where integral transform is to be evaluated (eg unprimed coordinates).
+    src_data : dict
+        Dictionary of data at source points. Keys should be those required by
+        kernel as kernel.keys. Vector data should be in rpz basis.
+    src_grid : LinearGrid
+        Source points for integral (eg primed coordinates). Should be linearly spaced
+        rectangular grid in both theta, zeta.
+    kernel : str or callable
+        Kernel function to evaluate. Should take 3 arguments:
+            eval_data : dict of data at evaluation points
+            src_data : dict of data at source points
+            diag : boolean, whether to evaluate full cross interactions or just diagonal
+        If a callable, should also have the attributes ``ndim`` and ``keys`` defined.
+        ``ndim`` is an integer representing the dimensionality of the output function f,
+        1 if f is scalar, 3 if f is a vector, etc.
+        ``keys`` is a list of strings of what data is required to evaluate the kernel.
+        The kernel will be called with dictionaries containing this data at source and
+        evaluation points.
+        If vector valued, the input to the kernel function will be in rpz and output
+        should be in xyz.
+    interpolator : callable
```

Comment:
these themselves have their own grids, which I assume should be the same as the `eval_grid` and `src_grid` being  passed into the singular integral call. Should add a check that the interpolator grids are the same as the passed-in grids

Reply:
Thats a good point, we can actually avoid having to pass in the grids directly themselves